File: iphone/Woladen/Services/FavoritesStore.swift
import Foundation

let favoriteCategoryUncategorized = "__uncategorized__"
let favoriteFilterAll = "__all__"

private let favoriteMetadataVersion = 2
private let maxFavoriteCategoryLength = 48
private let maxFavoriteCategories = 12

struct FavoriteItem: Codable, Equatable {
    var stationID: String
    var categories: [String]
    var createdAt: String
    var updatedAt: String
    var source: String

    enum CodingKeys: String, CodingKey {
        case stationID = "station_id"
        case categories
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case source
    }
}

private struct FavoriteMetadataPayload: Codable {
    let version: Int
    let items: [String: FavoriteItem]
}

@MainActor
final class FavoritesStore: ObservableObject {
    @Published private(set) var favorites: Set<String> = []
    @Published private(set) var items: [String: FavoriteItem] = [:]

    private let defaultsKey = "woladen_favorites_v2"
    private let legacyDefaultsKey = "woladen_favorites"
    private let seededFavorites: Set<String>?
    private var needsDeferredMigrationSave = false

    init() {
        seededFavorites = FavoritesStore.resolveSeededFavorites()
        if let seededFavorites {
            let timestamp = Self.nowISO8601()
            favorites = seededFavorites
            items = Dictionary(
                uniqueKeysWithValues: seededFavorites.map { stationID in
                    return (
                        stationID,
                        FavoriteItem(
                            stationID: stationID,
                            categories: [],
                            createdAt: timestamp,
                            updatedAt: timestamp,
                            source: "migration"
                        )
                    )
                }
            )
        } else {
            load()
            scheduleDeferredMigrationSaveIfNeeded()
        }
    }

    func toggle(_ stationID: String) {
        let id = normalizedStationID(stationID)
        guard !id.isEmpty else { return }
        if favorites.contains(id) {
            items.removeValue(forKey: id)
        } else {
            _ = ensureFavoriteItem(id, source: "manual")
        }
        syncFavorites()
        save()
    }

    func remove(_ stationID: String) {
        let id = normalizedStationID(stationID)
        guard favorites.contains(id) else { return }
        items.removeValue(forKey: id)
        syncFavorites()
        save()
    }

    func isFavorite(_ stationID: String) -> Bool {
        favorites.contains(normalizedStationID(stationID))
    }

    func categories(for stationID: String) -> [String] {
        items[normalizedStationID(stationID)]?.categories ?? []
    }

    func sortedCategories() -> [String] {
        var categories: [String] = []
        var seen: Set<String> = []
        for item in items.values {
            for category in normalizeCategories(item.categories, knownCategories: categories) {
                let key = categoryKey(category)
                guard seen.insert(key).inserted else { continue }
                categories.append(category)
            }
        }
        return categories.sorted { $0.localizedCaseInsensitiveCompare($1) == .orderedAscending }
    }

    func addCategory(_ category: String, to stationID: String, source: String = "manual") {
        let id = normalizedStationID(stationID)
        guard !id.isEmpty else { return }
        let label = normalizeCategoryLabel(category)
        guard !label.isEmpty else { return }

        let knownCategories = sortedCategories()
        var item = ensureFavoriteItem(id, source: source)
        item.categories = normalizeCategories(item.categories + [label], knownCategories: knownCategories)
        item.updatedAt = Self.nowISO8601()
        item.source = normalizedSource(item.source, fallback: source)
        items[id] = item
        syncFavorites()
        save()
    }

    func removeCategory(_ category: String, from stationID: String) {
        let id = normalizedStationID(stationID)
        let key = categoryKey(category)
        guard var item = items[id], !key.isEmpty else { return }
        item.categories.removeAll { categoryKey($0) == key }
        item.updatedAt = Self.nowISO8601()
        items[id] = item
        save()
    }

    func addRouteFavorites(stationIDs: [String], category: String) {
        let label = normalizeCategoryLabel(category)
        guard !label.isEmpty else { return }
        let knownCategories = sortedCategories()
        for stationID in stationIDs {
            let id = normalizedStationID(stationID)
            guard !id.isEmpty else { continue }
            var item = ensureFavoriteItem(id, source: "route")
            item.categories = normalizeCategories(item.categories + [label], knownCategories: knownCategories)
            item.updatedAt = Self.nowISO8601()
            item.source = normalizedSource(item.source, fallback: "route")
            items[id] = item
        }
        syncFavorites()
        save()
    }

    func categorySuggestions(query: String, excluding excluded: [String]) -> [String] {
        let excludedKeys = Set(excluded.map(categoryKey))
        let normalizedQuery = categoryKey(query)
        let categories = sortedCategories().filter { !excludedKeys.contains(categoryKey($0)) }
        if normalizedQuery.isEmpty {
            return Array(categories.prefix(6))
        }
        let prefix = categories.filter { categoryKey($0).hasPrefix(normalizedQuery) }
        let substring = categories.filter {
            let key = categoryKey($0)
            return !key.hasPrefix(normalizedQuery) && key.contains(normalizedQuery)
        }
        return Array((prefix + substring).prefix(6))
    }

    func favoriteIDs(matching filter: String) -> Set<String> {
        if filter == favoriteFilterAll {
            return favorites
        }
        if filter == favoriteCategoryUncategorized {
            return Set(items.values.filter { $0.categories.isEmpty }.map(\.stationID))
        }
        return Set(
            items.values
                .filter { item in item.categories.contains { categoryKey($0) == filter } }
                .map(\.stationID)
        )
    }

    func count(matching filter: String) -> Int {
        favoriteIDs(matching: filter).count
    }

    private func load() {
        guard seededFavorites == nil else { return }
        let defaults = UserDefaults.standard
        if let data = defaults.data(forKey: defaultsKey),
           let payload = try? JSONDecoder().decode(FavoriteMetadataPayload.self, from: data),
           payload.version == favoriteMetadataVersion {
            items = normalizeItems(payload.items)
            syncFavorites()
            return
        }

        let raw = defaults.array(forKey: legacyDefaultsKey) as? [String] ?? []
        let timestamp = Self.nowISO8601()
        items = Dictionary(
            uniqueKeysWithValues: Set(raw.map(normalizedStationID).filter { !$0.isEmpty }).map { stationID in
                (
                    stationID,
                    FavoriteItem(
                        stationID: stationID,
                        categories: [],
                        createdAt: timestamp,
                        updatedAt: timestamp,
                        source: "migration"
                    )
                )
            }
        )
        syncFavorites()
        needsDeferredMigrationSave = true
    }

    private func save() {
        guard seededFavorites == nil else { return }
        let knownCategories = sortedCategories()
        let payload = FavoriteMetadataPayload(
            version: favoriteMetadataVersion,
            items: Dictionary(
                uniqueKeysWithValues: items
                    .sorted { $0.key < $1.key }
                    .map { key, item in
                        (
                            key,
                            FavoriteItem(
                                stationID: key,
                                categories: normalizeCategories(item.categories, knownCategories: knownCategories),
                                createdAt: item.createdAt,
                                updatedAt: item.updatedAt,
                                source: normalizedSource(item.source)
                            )
                        )
                    }
            )
        )
        guard let data = try? JSONEncoder().encode(payload) else { return }
        UserDefaults.standard.set(data, forKey: defaultsKey)
    }

    private func ensureFavoriteItem(_ stationID: String, source: String) -> FavoriteItem {
        if let existing = items[stationID] {
            return existing
        }
        let timestamp = Self.nowISO8601()
        let item = FavoriteItem(
            stationID: stationID,
            categories: [],
            createdAt: timestamp,
            updatedAt: timestamp,
            source: normalizedSource(source)
        )
        items[stationID] = item
        return item
    }

    private func syncFavorites() {
        favorites = Set(items.keys)
    }

    private func scheduleDeferredMigrationSaveIfNeeded() {
        guard needsDeferredMigrationSave else { return }
        Task { @MainActor [weak self] in
            try? await Task.sleep(nanoseconds: 1_500_000_000)
            guard let self, self.needsDeferredMigrationSave else { return }
            self.needsDeferredMigrationSave = false
            self.save()
        }
    }

    private func normalizeItems(_ source: [String: FavoriteItem]) -> [String: FavoriteItem] {
        var normalized: [String: FavoriteItem] = [:]
        var knownCategories: [String] = []
        let fallbackTimestamp = Self.nowISO8601()
        for item in source.values {
            let stationID = normalizedStationID(item.stationID)
            guard !stationID.isEmpty else { continue }
            let categories = normalizeCategories(item.categories, knownCategories: knownCategories)
            knownCategories = normalizeCategories(knownCategories + categories)
            normalized[stationID] = FavoriteItem(
                stationID: stationID,
                categories: categories,
                createdAt: cleanTimestamp(item.createdAt, fallback: fallbackTimestamp),
                updatedAt: cleanTimestamp(item.updatedAt, fallback: item.createdAt),
                source: normalizedSource(item.source)
            )
        }
        return normalized
    }

    private static func resolveSeededFavorites() -> Set<String>? {
        let environment = ProcessInfo.processInfo.environment
        guard let rawFavorites = environment["WOLADEN_SCREENSHOT_FAVORITES"] else {
            return nil
        }

        let favorites = rawFavorites
            .split(separator: ",")
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }

        return Set(favorites)
    }

    private static func nowISO8601() -> String {
        ISO8601DateFormatter().string(from: Date())
    }
}

func normalizeCategoryLabel(_ value: String) -> String {
    String(value.replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression)
        .trimmingCharacters(in: .whitespacesAndNewlines)
        .prefix(maxFavoriteCategoryLength))
}

func categoryKey(_ value: String) -> String {
    normalizeCategoryLabel(value).lowercased()
}

private func normalizedStationID(_ value: String) -> String {
    value.trimmingCharacters(in: .whitespacesAndNewlines)
}

private func normalizeCategories(_ value: [String], knownCategories: [String] = []) -> [String] {
    var displayByKey: [String: String] = [:]
    for category in knownCategories {
        let label = normalizeCategoryLabel(category)
        let key = categoryKey(label)
        if !label.isEmpty, displayByKey[key] == nil {
            displayByKey[key] = label
        }
    }

    var categories: [String] = []
    var seen: Set<String> = []
    for item in value {
        let label = normalizeCategoryLabel(item)
        let key = categoryKey(label)
        guard !label.isEmpty, seen.insert(key).inserted else { continue }
        categories.append(displayByKey[key] ?? label)
    }
    return Array(categories.prefix(maxFavoriteCategories))
}

private func normalizedSource(_ value: String, fallback: String = "manual") -> String {
    let source = value.trimmingCharacters(in: .whitespacesAndNewlines)
    return ["manual", "route", "migration"].contains(source) ? source : fallback
}

private func cleanTimestamp(_ value: String, fallback: String) -> String {
    let text = value.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !text.isEmpty else { return fallback }
    let formatter = ISO8601DateFormatter()
    return formatter.date(from: text) == nil ? fallback : text
}

File: android/app/src/main/java/de/woladen/android/store/FavoritesStore.kt
package de.woladen.android.store

import android.content.Context
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import org.json.JSONObject
import java.time.Instant
import java.util.Locale

const val FAVORITE_CATEGORY_UNCATEGORIZED = "__uncategorized__"
const val FAVORITE_FILTER_ALL = "__all__"

private const val FAVORITE_METADATA_VERSION = 2
private const val MAX_FAVORITE_CATEGORY_LENGTH = 48
private const val MAX_FAVORITE_CATEGORIES = 12

data class FavoriteItem(
    val stationId: String,
    val categories: List<String>,
    val createdAt: String,
    val updatedAt: String,
    val source: String
)

class FavoritesStore(context: Context) {
    private val preferences = context.getSharedPreferences("woladen", Context.MODE_PRIVATE)
    private val defaultsKey = "woladen_favorites_v2"
    private val legacyDefaultsKey = "woladen_favorites"

    var items: Map<String, FavoriteItem> by mutableStateOf(loadItems())
        private set

    var favorites: Set<String> by mutableStateOf(items.keys)
        private set

    fun toggle(stationId: String) {
        val id = normalizeStationId(stationId)
        if (id.isBlank()) return
        val next = items.toMutableMap()
        if (next.containsKey(id)) {
            next.remove(id)
        } else {
            next[id] = createFavoriteItem(id, source = "manual")
        }
        publish(next)
        saveItems()
    }

    fun remove(stationId: String) {
        val id = normalizeStationId(stationId)
        if (!favorites.contains(id)) return
        val next = items.toMutableMap()
        next.remove(id)
        publish(next)
        saveItems()
    }

    fun isFavorite(stationId: String): Boolean {
        return favorites.contains(normalizeStationId(stationId))
    }

    fun categoriesFor(stationId: String): List<String> {
        return items[normalizeStationId(stationId)]?.categories.orEmpty()
    }

    fun sortedCategories(): List<String> {
        val categories = mutableListOf<String>()
        val seen = linkedSetOf<String>()
        for (item in items.values) {
            for (category in normalizeCategories(item.categories, knownCategories = categories)) {
                val key = categoryKey(category)
                if (seen.add(key)) {
                    categories += category
                }
            }
        }
        return categories.sortedWith(String.CASE_INSENSITIVE_ORDER)
    }

    fun addCategory(stationId: String, category: String, source: String = "manual") {
        val id = normalizeStationId(stationId)
        val label = normalizeCategoryLabel(category)
        if (id.isBlank() || label.isBlank()) return
        val next = items.toMutableMap()
        val existing = next[id] ?: createFavoriteItem(id, source)
        val knownCategories = sortedCategories()
        next[id] = existing.copy(
            categories = normalizeCategories(existing.categories + label, knownCategories = knownCategories),
            updatedAt = nowIso(),
            source = normalizeSource(existing.source, fallback = source)
        )
        publish(next)
        saveItems()
    }

    fun removeCategory(stationId: String, category: String) {
        val id = normalizeStationId(stationId)
        val key = categoryKey(category)
        val existing = items[id] ?: return
        val next = items.toMutableMap()
        next[id] = existing.copy(
            categories = existing.categories.filter { categoryKey(it) != key },
            updatedAt = nowIso()
        )
        publish(next)
        saveItems()
    }

    fun addRouteFavorites(stationIds: List<String>, category: String) {
        val label = normalizeCategoryLabel(category)
        if (label.isBlank()) return
        val next = items.toMutableMap()
        val knownCategories = sortedCategories()
        for (rawStationId in stationIds) {
            val stationId = normalizeStationId(rawStationId)
            if (stationId.isBlank()) continue
            val existing = next[stationId] ?: createFavoriteItem(stationId, source = "route")
            next[stationId] = existing.copy(
                categories = normalizeCategories(existing.categories + label, knownCategories = knownCategories),
                updatedAt = nowIso(),
                source = normalizeSource(existing.source, fallback = "route")
            )
        }
        publish(next)
        saveItems()
    }

    fun categorySuggestions(query: String, excluding: List<String>): List<String> {
        val excludedKeys = excluding.mapTo(linkedSetOf(), ::categoryKey)
        val normalizedQuery = categoryKey(query)
        val categories = sortedCategories().filter { categoryKey(it) !in excludedKeys }
        if (normalizedQuery.isBlank()) {
            return categories.take(6)
        }
        val prefix = categories.filter { categoryKey(it).startsWith(normalizedQuery) }
        val substring = categories.filter {
            val key = categoryKey(it)
            !key.startsWith(normalizedQuery) && key.contains(normalizedQuery)
        }
        return (prefix + substring).take(6)
    }

    fun favoriteIdsMatching(filter: String): Set<String> {
        return when (filter) {
            FAVORITE_FILTER_ALL -> favorites
            FAVORITE_CATEGORY_UNCATEGORIZED -> items.values
                .filter { it.categories.isEmpty() }
                .mapTo(linkedSetOf()) { it.stationId }
            else -> items.values
                .filter { item -> item.categories.any { categoryKey(it) == filter } }
                .mapTo(linkedSetOf()) { it.stationId }
        }
    }

    fun countMatching(filter: String): Int {
        return favoriteIdsMatching(filter).size
    }

    private fun loadItems(): Map<String, FavoriteItem> {
        preferences.getString(defaultsKey, null)?.let { raw ->
            parseItems(raw)?.let { return it }
        }

        val timestamp = nowIso()
        return preferences.getStringSet(legacyDefaultsKey, emptySet())
            .orEmpty()
            .map(::normalizeStationId)
            .filter { it.isNotBlank() }
            .distinct()
            .associateWith { stationId ->
                FavoriteItem(
                    stationId = stationId,
                    categories = emptyList(),
                    createdAt = timestamp,
                    updatedAt = timestamp,
                    source = "migration"
                )
            }
    }

    private fun saveItems() {
        val payload = JSONObject()
        payload.put("version", FAVORITE_METADATA_VERSION)
        val jsonItems = JSONObject()
        val knownCategories = sortedCategories()
        for ((stationId, item) in items.toSortedMap()) {
            jsonItems.put(stationId, JSONObject().apply {
                put("station_id", stationId)
                put("categories", org.json.JSONArray(normalizeCategories(item.categories, knownCategories)))
                put("created_at", item.createdAt)
                put("updated_at", item.updatedAt)
                put("source", normalizeSource(item.source))
            })
        }
        payload.put("items", jsonItems)
        preferences.edit().putString(defaultsKey, payload.toString()).apply()
    }

    private fun publish(next: Map<String, FavoriteItem>) {
        items = next.toMap()
        favorites = items.keys
    }

    private fun createFavoriteItem(stationId: String, source: String): FavoriteItem {
        val timestamp = nowIso()
        return FavoriteItem(
            stationId = stationId,
            categories = emptyList(),
            createdAt = timestamp,
            updatedAt = timestamp,
            source = normalizeSource(source)
        )
    }

    private fun parseItems(raw: String): Map<String, FavoriteItem>? {
        return runCatching {
            val payload = JSONObject(raw)
            if (payload.optInt("version") != FAVORITE_METADATA_VERSION) {
                return@runCatching null
            }
            val sourceItems = payload.optJSONObject("items") ?: return@runCatching emptyMap()
            val result = linkedMapOf<String, FavoriteItem>()
            val knownCategories = mutableListOf<String>()
            val keys = sourceItems.keys()
            while (keys.hasNext()) {
                val key = keys.next()
                val item = sourceItems.optJSONObject(key) ?: continue
                val stationId = normalizeStationId(item.optString("station_id", key))
                if (stationId.isBlank()) continue
                val categories = normalizeCategories(
                    (0 until item.optJSONArray("categories")?.length().orZero())
                        .mapNotNull { index -> item.optJSONArray("categories")?.optString(index) },
                    knownCategories = knownCategories
                )
                knownCategories += categories
                val createdAt = cleanTimestamp(item.optString("created_at"), fallback = nowIso())
                result[stationId] = FavoriteItem(
                    stationId = stationId,
                    categories = categories,
                    createdAt = createdAt,
                    updatedAt = cleanTimestamp(item.optString("updated_at"), fallback = createdAt),
                    source = normalizeSource(item.optString("source"))
                )
            }
            result
        }.getOrNull()
    }
}

fun normalizeCategoryLabel(value: String): String {
    return value
        .replace("\\s+".toRegex(), " ")
        .trim()
        .take(MAX_FAVORITE_CATEGORY_LENGTH)
}

fun categoryKey(value: String): String {
    return normalizeCategoryLabel(value).lowercase(Locale.ROOT)
}

private fun normalizeStationId(value: String): String = value.trim()

private fun normalizeCategories(value: List<String>, knownCategories: List<String> = emptyList()): List<String> {
    val displayByKey = linkedMapOf<String, String>()
    for (category in knownCategories) {
        val label = normalizeCategoryLabel(category)
        val key = categoryKey(label)
        if (label.isNotBlank() && key !in displayByKey) {
            displayByKey[key] = label
        }
    }

    val categories = mutableListOf<String>()
    val seen = linkedSetOf<String>()
    for (item in value) {
        val label = normalizeCategoryLabel(item)
        val key = categoryKey(label)
        if (label.isBlank() || !seen.add(key)) continue
        categories += displayByKey[key] ?: label
    }
    return categories.take(MAX_FAVORITE_CATEGORIES)
}

private fun normalizeSource(value: String, fallback: String = "manual"): String {
    val source = value.trim()
    return if (source in setOf("manual", "route", "migration")) source else fallback
}

private fun cleanTimestamp(value: String, fallback: String): String {
    val text = value.trim()
    if (text.isBlank()) return fallback
    return runCatching { Instant.parse(text) }.fold(
        onSuccess = { text },
        onFailure = { fallback }
    )
}

private fun nowIso(): String = Instant.now().toString()

private fun Int?.orZero(): Int = this ?: 0
